src/json-patch.ts

```diff
@@ -257,30 +270,120 @@ module jsonpatch {
+   * Apply a single JSON Patch Operation on a JSON document.
+   * Returns {newDocument, result} of the operation.
+   *
+   * @param document The document to patch
+   * @param operation The operation to apply
+   * @param validate Whether to validate the operation
+   * @param mutateDocument Whether to mutate the original document or clone it before applying
+   * @return `{newDocument, result}` after the operation
    */
-  export function apply(tree: any, patches: Patch<any>[], validate?: boolean): any[] {
-    var results = new Array(patches.length)
-      , p = 0
-      , plen = patches.length
-      , patch
-      , key;
-    while (p < plen) {
-      patch = patches[p];
-      p++;
-      // Find the object
-      var path = patch.path || "";
-      var keys = path.split('/');
-      var obj = tree;
-      var t = 1; //skip empty element - http://jsperf.com/to-shift-or-not-to-shift
-      var len = keys.length;
-      var existingPathFragment = undefined;
-
-      while (true) {
+  export function applyOperation<T>(document: T, operation: Operation): OperationResult;
+
+  /**
+   * Apply a single JSON Patch Operation on a JSON document.
+   * Returns the {newDocument, result} of the operation.
+   *
+   * @param document The document to patch
+   * @param operation The operation to apply
+   * @param validate Whether to validate the operation
+   * @param mutateDocument Whether to mutate the original document or clone it before applying
+   * @return `{newDocument, result}` after the operation
+   */
+  export function applyOperation<T>(document: any, operation: Operation, validate: boolean, mutateDocument: boolean): OperationResult;
+  export function applyOperation<T>(document: any, operation: Operation, validate = false, mutateDocument = true): OperationResult {
+    if (validate) {
+      this.validator(operation, 0);
+    }
+    const returnValue: OperationResult = { newDocument: document, result: undefined };
+    /* ROOT OPERATIONS */
+    if (operation.path === "") {
+      if (operation.op === 'add') {
+        returnValue.newDocument = operation.value;
+        returnValue.result = undefined;
+        return returnValue;
+      } else if (operation.op === 'replace') {
+        returnValue.newDocument = operation.value;
+        returnValue.result = document; //document we removed
+        return returnValue;
+      }
+      else if (operation.op === 'move' || operation.op === 'copy') { // it's a move or copy to root
+        returnValue.newDocument = jsonpatch.getValueByPointer(document, operation.from); // get the value by json-pointer in `from` field
+        if (operation.op === 'move') { // report removed item
+          returnValue.result = document;
+        }
+        return returnValue;
+      } else { // it's a remove or test on root
+        if (operation.op === 'test') {
+          returnValue.result = _equals(document, operation.value);
+          returnValue.newDocument = document;
+          return returnValue;
```

Comment:
It was never ensured. And I guess the only possible solution is to deep clone when there is a `test` op. And not here, but in `applyPatch`.